src/interpreter.ts

```diff
@@ -25,3 +499,840 @@ export function parseAndEvalExpression(sourceCode: string): EvalResult {
+        // which is `-(2**256)` we need to have a special case for it
+
+        if (operand.kind === "number" && op === "-") {
+            // emulating negative integer literals
+            return ensureInt(-operand.value, source);
+        }
+
+        const valOperand = this.interpretExpression(operand);
+
+        return evalUnaryOp(op, valOperand, operand.loc, source);
+    }
+
+    public interpretBinaryOp(ast: AstOpBinary): Value {
+        const left = ast.left;
+        const right = ast.right;
+        const op = ast.op;
+        const source = ast.loc;
+
+        const valLeft = this.interpretExpression(left);
+        const valRight = this.interpretExpression(right);
+
+        return evalBinaryOp(op, valLeft, valRight, left.loc, right.loc, source);
+    }
+
+    public interpretConditional(ast: AstConditional): Value {
+        const condition = ast.condition;
+        const thenBranch = ast.thenBranch;
+        const elseBranch = ast.elseBranch;
+
+        // here we rely on the typechecker that both branches have the same type
+        const valCond = ensureBoolean(
+            this.interpretExpression(condition),
+            condition.loc,
+        );
+        if (valCond) {
+            return this.interpretExpression(thenBranch);
+        } else {
+            return this.interpretExpression(elseBranch);
+        }
+    }
+
+    public interpretStructInstance(ast: AstStructInstance): StructValue {
+        const structTypeId = ast.type;
+        const structFields = ast.args;
+
+        const structTy = getType(this.initialContext, structTypeId);
+
+        // initialize the resulting struct value with
+        // the default values for fields with initializers
+        // or null for uninitialized optional fields
+        const resultWithDefaultFields: StructValue = structTy.fields.reduce(
+            (resObj, field) => {
+                if (field.default !== undefined) {
+                    resObj[field.name] = field.default;
+                } else {
+                    if (field.type.kind === "ref" && field.type.optional) {
+                        resObj[field.name] = null;
+                    }
+                }
+                return resObj;
+            },
+            { $tactStruct: idText(structTypeId) } as StructValue,
+        );
+
+        // this will override default fields set above
+        return structFields.reduce((resObj, fieldWithInit) => {
+            resObj[idText(fieldWithInit.field)] = this.interpretExpression(
+                fieldWithInit.initializer,
+            );
+            return resObj;
+        }, resultWithDefaultFields);
+    }
+
+    public interpretFieldAccess(ast: AstFieldAccess): Value {
+        const structExpr = ast.aggregate;
+        const fieldId = ast.field;
+
+        // special case for contract/trait constant accesses via `self.constant`
+        if (structExpr.kind === "id" && isSelfId(structExpr)) {
```

Comment:
A struct can also be referred to as `self` if it's the self argument of a method. Looks like you need to filter out this case, otherwise a legitimate case would generate and error